x/httpproxy/connect_client.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2023 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpproxy
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"net"
+	"net/http"
+)
+
+// connectClient is a [transport.StreamDialer] implementation that sends a CONNECT request
+type connectClient struct {
+	endpoint transport.StreamEndpoint
+
+	// proxyAuth is the Proxy-Authorization header value. If empty, the header is not sent
+	proxyAuth string
+}
+
+var _ transport.StreamDialer = (*connectClient)(nil)
+
+type ConnectClientOption func(c *connectClient)
+
+func NewConnectClient(endpoint transport.StreamEndpoint, opts ...ConnectClientOption) (transport.StreamDialer, error) {
+	if endpoint == nil {
+		return nil, errors.New("endpoint must not be nil")
+	}
+
+	cc := &connectClient{
+		endpoint: endpoint,
+	}
+
+	for _, opt := range opts {
+		opt(cc)
+	}
+
+	return cc, nil
+}
+
+// WithProxyAuthorization - sets the Proxy-Authorization header value
+func WithProxyAuthorization(proxyAuth string) ConnectClientOption {
+	return func(c *connectClient) {
+		c.proxyAuth = proxyAuth
+	}
+}
+
+// DialStream - connects using the endpoint and sends a CONNECT request to the remoteAddr, closes the connection if the request fails
+func (c *connectClient) DialStream(ctx context.Context, remoteAddr string) (transport.StreamConn, error) {
+	conn, err := c.endpoint.ConnectStream(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to endpoint: %w", err)
+	}
+
+	err = c.sendConnectRequest(ctx, remoteAddr, conn)
+	if err != nil {
+		conn.Close()
+		return nil, fmt.Errorf("sendConnectRequest: %w", err)
+	}
+
+	return conn, nil
+}
+
+func (c *connectClient) sendConnectRequest(ctx context.Context, remoteAddr string, conn transport.StreamConn) error {
```

Comment:
I was hoping to use the `http.Client` to avoid replicating logic, provide extra functionality, and keep it less tied to HTTP/1.1.

Does https://pkg.go.dev/net/http#Client.Do not work with a CONNECT method?

I think you may need https://pkg.go.dev/net/http#NewResponseController to enable full duplex and hijack the connection.